homeassistant/components/essent/sensor.py

```diff
@@ -0,0 +1,101 @@
+"""Support for Essent API."""
+
+import xml.etree.ElementTree as ET
+
+from pyessent import PyEssent
+
+from homeassistant.const import ENERGY_KILO_WATT_HOUR, STATE_UNKNOWN
+from homeassistant.helpers.entity import Entity
+
+DOMAIN = 'essent'
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Essent platform."""
+    username = config['username']
+    password = config['password']
+    for meter in EssentBase(username, password).retrieve_meters():
+        for tariff in ['L', 'N', 'H']:
+            add_devices([EssentMeter(username, password, meter, tariff)])
+
+
+class EssentBase():
+    """Essent Base."""
+
+    def __init__(self, username, password):
+        """Initialize the Essent API."""
+        self._essent = PyEssent(username, password)
+
+    def get_session(self):
+        """Return the active session."""
+        return self._essent
+
+    def retrieve_meters(self):
+        """Retrieve the IDs of the meters used by Essent."""
+        return self._essent.get_EANs()
+
+
+class EssentMeter(Entity):
+    """Representation of Essent measurements."""
+
+    def __init__(self, username, password, meter, tariff):
+        """Initialize the sensor."""
+        self._state = None
+        self._username = username
+        self._password = password
+        self._meter = meter
+        self._tariff = tariff
+        self._meter_type = STATE_UNKNOWN
+        self._meter_unit = None
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return "Essent {} ({})".format(self._meter_type, self._tariff)
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement."""
+        if self._meter_unit and self._meter_unit.lower() == 'kwh':
+            return ENERGY_KILO_WATT_HOUR
+
+        return self._meter_unit
+
+    def update(self):
+        """Fetch the energy usage."""
+        # Retrieve an authenticated session
+        essent = EssentBase(self._username, self._password).get_session()
+
+        # Read the meter
+        meter_request = essent.Customer.get_meter_reading_history(
+            self._meter, only_last_meter_reading=True)
+
+        # Parse out into the root of our data
+        info_base = ET.fromstring(meter_request.text) \
```

Comment:
I know I'm sounding pedantic, but why can't ALL xml parsing stuff be moved to PyEssent ?

Reply:
This part is pretty specific to how to display it all though. It would seem to me that parsing the data in the best format to display it is more Home Assistant's responsibility and moving this into PyEssent would just make it harder for others to use PyEssent for anything useful.

Reply:
One of the reasons we ask every communication to be moved into an external library is that more often than not API's break and we end up going through a lengthy review process (updating an external library version is usually quicker).

Reply:
Hmm, I understand your point. I'll sleep over it and see if I can think of a way to move as much as possible to PyEssent without making that library too Home Assistant-specific. Try to figure out the most "neutral" way I can implement this in PyEssent.